File: p1.py
#Time: O(2n)
#Space: O(n)
class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:

        if T is None or len(T)==0:
            return []

        stack = []
        result = [0]*len(T)
        for i in range(len(T)):
            while stack and T[i] > T[stack[-1]]:
                idx = stack.pop()
                result[idx] = i - idx

            stack.append(i)

        return result

                

File: p1.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach


class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> stk = new Stack<>();
        int[] result = new int[temperatures.length];
        //Go through all the elements
        for(int i=0; i<temperatures.length; i++){
            //If stack is empty or the element in stack is greater or equal than the current elemnt, we put current in stack
            if(stk.isEmpty() || temperatures[stk.peek()] >= temperatures[i]){
                stk.push(i);
            }
            //If we find the elemnt greater than the element in stack, we calculate the number of days
            else{
                while(!stk.isEmpty() && temperatures[stk.peek()] < temperatures[i]){
                    int j = stk.pop();
                    result[j] = i-j;
                }
                stk.push(i);
            }
        }

        return result;
    }
}